gcc/rust/util/rust-unicode.cc

```diff
@@ -0,0 +1,328 @@
+      mid = (low + high) / 2;
+    }
+
+  if (array[mid] == target)
+    return mid;
+  else
+    return -1;
+}
+
+int
+lookup_cc (codepoint_t c)
+{
+  auto it = Rust::CCC_TABLE.find (c);
+  if (it != Rust::CCC_TABLE.end ())
+    return it->second;
+  else
+    // Starter. Returns zero.
+    return 0;
+}
+
+tl::optional<codepoint_t>
+lookup_recomp (codepoint_t starter, codepoint_t c)
+{
+  auto it = Rust::RECOMPOSITION_MAP.find ({starter, c});
+  if (it != Rust::RECOMPOSITION_MAP.end ())
+    return {it->second};
+
+  it = Rust::RECOMPOSITION_MAP.find ({starter, 0});
+  if (it != Rust::RECOMPOSITION_MAP.end ())
+    return {it->second};
+
+  return tl::nullopt;
+}
+
+void
+recursive_decomp_cano (codepoint_t c, string_t &buf)
+{
+  auto it = Rust::DECOMPOSITION_MAP.find (c);
+  if (it != Rust::DECOMPOSITION_MAP.end ())
+    {
+      string_t decomped = it->second;
+      for (codepoint_t cp : decomped)
+	{
+	  recursive_decomp_cano (cp, buf);
+	}
+    }
+  else
+    buf.push_back (c);
+}
+
+string_t
+decomp_cano (string_t s)
+{
+  // TODO: Algorithmic lookup for Hangul
+  string_t buf;
+  for (codepoint_t c : s)
+    recursive_decomp_cano (c, buf);
+  return buf;
+}
+
+void
+sort_cano (string_t &s)
+{
+  int cc_here, cc_prev;
+  if (s.size () == 1)
+    return;
+  for (unsigned int i = 1; i < s.size (); i++)
+    {
+      cc_here = lookup_cc (s[i]);
+      cc_prev = lookup_cc (s[i - 1]);
+      if (cc_here >= 0 && cc_prev > cc_here)
+	{
+	  // swap
+	  int tmp = s[i];
+	  s[i] = s[i - 1];
+	  s[i - 1] = tmp;
+	  if (i > 1)
+	    i -= 2;
+	}
+    }
+}
+
+string_t
+recomp (string_t s)
+{
+  // TODO: Algorithmic lookup for Hangul
+  string_t buf;
+  if (s.size () > 0)
+    {
+      int last_class = -1;
+      // int starter_pos = 0; // Assume the first character is Starter. Correct?
+      // int target_pos = 1;
+      codepoint_t starter_ch = s[0];
+      for (unsigned int src_pos = 1; src_pos < s.size (); src_pos++)
+	{
+	  // get current character
+	  codepoint_t ch = s[src_pos];
+	  int ch_class = lookup_cc (ch);
+	  tl::optional<codepoint_t> composite = lookup_recomp (starter_ch, ch);
+	  if (composite.has_value () && last_class < ch_class)
+	    {
+	      // ch can be composed
+	      buf.push_back (composite.value ());
+	      starter_ch = composite.value ();
+	    }
+	  else if (ch_class == 0)
+	    {
+	      // ch is Starter and cannot be composed.
+	      if (src_pos == 1)
+		// FIXME: buggy?
+		buf.push_back (starter_ch);
+	      // starter_pos = target_pos;
+	      starter_ch = ch;
+	      last_class = -1;
+	      buf.push_back (ch);
+	    }
+	  else
+	    {
+	      // ch is not Starter.
+	      last_class = ch_class;
+	      buf.push_back (ch);
+	    }
+	}
+    }
+  return buf;
+}
+
+// TODO: remove
+/*
+void
+dump_string (std::vector<uint32_t> s)
+{
+  std::cout << "dump=";
+  for (auto c : s)
+    {
+      std::cout << std::hex << c << ", ";
+    }
+  std::cout << std::endl;
+}
+*/
+
+string_t
+nfc_normalize (string_t s)
+{
```

Comment:
This function performs Unicode normalization of the given string.
Currently `string_t` aliases std::vector<uint32_t>`, but it will be replaced with the class `Utf8String`, introduced via PR https://github.com/Rust-GCC/gccrs/pull/2463